File: src/core/project-manager.ts
import fs from 'fs/promises';
import path from 'path';
import { configStore } from '@core/config-store';

export interface ProjectMetadata {
  id: string;
  name: string;
  createdAt: string;
  updatedAt: string;
  coverUrl?: string;
  description?: string;
  videoUrl?: string;
}

export interface ProjectIndex {
  projects: ProjectMetadata[];
}

function resolveWorkspaceRoot(): string {
  const envRoot = process.env.MANGOU_HOME;
  if (envRoot && envRoot.trim()) {
    return path.resolve(envRoot.trim());
  }
  const explicitProjectsRoot = process.env.MANGOU_WORKSPACE_ROOT;
  if (explicitProjectsRoot && explicitProjectsRoot.trim()) {
    return path.dirname(path.resolve(explicitProjectsRoot.trim()));
  }
  return process.cwd();
}

function resolveProjectsRoot(workspaceRoot: string): string {
  const explicitProjectsRoot = process.env.MANGOU_WORKSPACE_ROOT;
  if (explicitProjectsRoot && explicitProjectsRoot.trim()) {
    return path.resolve(explicitProjectsRoot.trim());
  }
  return path.resolve(workspaceRoot, configStore.get('workspaceDir'));
}

async function fileExists(targetPath: string): Promise<boolean> {
  try {
    await fs.access(targetPath);
    return true;
  } catch {
    return false;
  }
}

export class ProjectManager {
  private static defaultManager: ProjectManager | null = null;
  private readonly workspaceRoot: string;
  private readonly projectsRoot: string;
  private readonly indexPath: string;

  constructor(options?: string | { workspaceRoot?: string; projectsRoot?: string }) {
    const workspaceRoot = typeof options === 'string' ? options : options?.workspaceRoot;
    const projectsRoot = typeof options === 'string' ? undefined : options?.projectsRoot;

    this.workspaceRoot = workspaceRoot ? path.resolve(workspaceRoot) : resolveWorkspaceRoot();
    this.projectsRoot = projectsRoot
      ? path.resolve(projectsRoot)
      : resolveProjectsRoot(this.workspaceRoot);
    this.indexPath = path.join(this.workspaceRoot, 'projects.json');
  }

  static getDefault(): ProjectManager {
    if (!this.defaultManager) {
      this.defaultManager = new ProjectManager();
    }
    return this.defaultManager;
  }

  static async listProjects(): Promise<ProjectMetadata[]> {
    const manager = ProjectManager.getDefault();
    await manager.init();
    return manager.listProjects();
  }

  static async getProject(id: string): Promise<ProjectMetadata | null> {
    const manager = ProjectManager.getDefault();
    await manager.init();
    return manager.getProject(id);
  }

  async init(): Promise<void> {
    await fs.mkdir(this.workspaceRoot, { recursive: true });
    await fs.mkdir(this.projectsRoot, { recursive: true });
    if (!(await fileExists(this.indexPath))) {
      await this.saveIndex({ projects: [] });
    }
  }

  async listProjects(): Promise<ProjectMetadata[]> {
    const indexed = await this.readIndex();
    const scanned = await this.scanProjectsRoot();
    const scannedById = new Map(scanned.map((project) => [project.id, project]));

    const merged = indexed.projects
      .map((project) => scannedById.get(project.id) || null)
      .filter((project): project is ProjectMetadata => !!project);

    for (const project of scanned) {
      if (!merged.some((item) => item.id === project.id)) {
        merged.push(project);
      }
    }

    await this.saveIndex({ projects: merged });

    return merged.sort(
      (a, b) => new Date(b.updatedAt).getTime() - new Date(a.updatedAt).getTime()
    );
  }

  async getProject(projectId: string): Promise<ProjectMetadata | null> {
    const projects = await this.listProjects();
    return projects.find((project) => project.id === projectId) || null;
  }

  async updateProjectVideoUrl(projectId: string, videoUrl: string): Promise<ProjectMetadata | null> {
    const projects = await this.listProjects();
    const project = projects.find((item) => item.id === projectId);
    if (!project) return null;

    const now = new Date().toISOString();
    project.videoUrl = videoUrl;
    project.updatedAt = now;

    const projectJsonPath = path.join(this.projectsRoot, projectId, 'project.json');
    let projectJson: Record<string, unknown> = {};
    try {
      projectJson = JSON.parse(await fs.readFile(projectJsonPath, 'utf-8'));
    } catch {}
    await fs.writeFile(
      projectJsonPath,
      `${JSON.stringify({ ...projectJson, videoUrl, updatedAt: now }, null, 2)}\n`,
      'utf-8'
    );

    await this.saveIndex({ projects });
    return project;
  }

  private async scanProjectsRoot(): Promise<ProjectMetadata[]> {
    const entries = await fs.readdir(this.projectsRoot, { withFileTypes: true }).catch(() => []);
    const projects: ProjectMetadata[] = [];

    for (const entry of entries) {
      if (!entry.isDirectory() || entry.name.startsWith('.')) continue;

      const projectId = entry.name;
      const projectRoot = path.join(this.projectsRoot, projectId);
      const storyboardsDir = path.join(projectRoot, 'storyboards');
      const tasksPath = path.join(projectRoot, 'tasks.jsonl');
      const projectJsonPath = path.join(projectRoot, 'project.json');

      if (!(await fileExists(storyboardsDir))) continue;
      if (!(await fileExists(tasksPath))) {
        await fs.writeFile(tasksPath, '', 'utf-8');
      }

      let projectJson: Record<string, any> = {};
      try {
        projectJson = JSON.parse(await fs.readFile(projectJsonPath, 'utf-8'));
      } catch {}

      const now = new Date().toISOString();
      const metadata: ProjectMetadata = {
        id: projectId,
        name: projectJson.name || projectId,
        description: projectJson.description || '',
        coverUrl: projectJson.coverUrl,
        videoUrl: projectJson.videoUrl,
        createdAt: projectJson.createdAt || now,
        updatedAt: projectJson.updatedAt || projectJson.createdAt || now,
      };

      projects.push(metadata);
    }

    return projects;
  }

  private async readIndex(): Promise<ProjectIndex> {
    try {
      const content = await fs.readFile(this.indexPath, 'utf-8');
      const parsed = JSON.parse(content);
      return Array.isArray(parsed?.projects) ? parsed : { projects: [] };
    } catch {
      return { projects: [] };
    }
  }

  private async saveIndex(index: ProjectIndex): Promise<void> {
    await fs.writeFile(this.indexPath, `${JSON.stringify(index, null, 2)}\n`, 'utf-8');
  }
}

File: mangou_skill/project_manager.py
from __future__ import annotations

import json
import os
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from .config import load_config
from .project import resolve_workspace_and_projects_root


@dataclass
class ProjectMetadata:
    id: str
    name: str
    createdAt: str
    updatedAt: str
    coverUrl: str | None = None
    description: str = ""
    videoUrl: str | None = None


def _resolve_workspace_root() -> Path:
    return resolve_workspace_and_projects_root()[0]


def _resolve_projects_root(workspace_root: Path) -> Path:
    if os.environ.get("MANGOU_HOME"):
        return workspace_root / load_config(workspace_root).workspace_dir
    return resolve_workspace_and_projects_root()[1]


class ProjectManager:
    def __init__(self, workspace_root: str | Path | None = None, projects_root: str | Path | None = None) -> None:
        self.workspace_root = Path(workspace_root).resolve() if workspace_root else _resolve_workspace_root()
        if projects_root:
            self.projects_root = Path(projects_root).resolve()
        elif workspace_root:
            self.projects_root = self.workspace_root / load_config(self.workspace_root).workspace_dir
        else:
            self.projects_root = _resolve_projects_root(self.workspace_root)
        self.index_path = self.workspace_root / "projects.json"

    def init(self) -> None:
        self.workspace_root.mkdir(parents=True, exist_ok=True)
        self.projects_root.mkdir(parents=True, exist_ok=True)
        if not self.index_path.exists():
            self._save_index({"projects": []})

    def list_projects(self) -> list[dict[str, object]]:
        indexed = self._read_index().get("projects", [])
        scanned = self._scan_projects_root()
        scanned_by_id = {project["id"]: project for project in scanned}
        merged = [scanned_by_id[project["id"]] for project in indexed if project.get("id") in scanned_by_id]
        merged_ids = {project["id"] for project in merged}
        for project in scanned:
            if project["id"] not in merged_ids:
                merged.append(project)
        merged.sort(key=lambda item: item.get("updatedAt", ""), reverse=True)
        self._save_index({"projects": merged})
        return merged

    def get_project(self, project_id: str) -> dict[str, object] | None:
        for project in self.list_projects():
            if project.get("id") == project_id:
                return project
        return None

    def _scan_projects_root(self) -> list[dict[str, object]]:
        projects: list[dict[str, object]] = []
        if not self.projects_root.exists():
            return projects

        for entry in self.projects_root.iterdir():
            if not entry.is_dir() or entry.name.startswith("."):
                continue
            storyboards_dir = entry / "storyboards"
            if not storyboards_dir.exists():
                continue
            tasks_path = entry / "tasks.jsonl"
            tasks_path.touch(exist_ok=True)

            project_json_path = entry / "project.json"
            try:
                project_json = json.loads(project_json_path.read_text(encoding="utf-8"))
            except Exception:
                project_json = {}

            now = datetime.now(timezone.utc).isoformat()
            projects.append(
                {
                    "id": entry.name,
                    "name": project_json.get("name") or entry.name,
                    "description": project_json.get("description") or "",
                    "coverUrl": project_json.get("coverUrl"),
                    "videoUrl": project_json.get("videoUrl"),
                    "createdAt": project_json.get("createdAt") or project_json.get("created_at") or now,
                    "updatedAt": project_json.get("updatedAt") or project_json.get("updated_at") or project_json.get("createdAt") or project_json.get("created_at") or now,
                }
            )
        return projects

    def _read_index(self) -> dict[str, list[dict[str, object]]]:
        try:
            parsed = json.loads(self.index_path.read_text(encoding="utf-8"))
        except Exception:
            return {"projects": []}
        if isinstance(parsed, dict) and isinstance(parsed.get("projects"), list):
            return parsed
        return {"projects": []}

    def _save_index(self, index: dict[str, list[dict[str, object]]]) -> None:
        self.index_path.write_text(json.dumps(index, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
